Enable Filter Stages to be completed asynchronously without re-ordering messages

robobario opened the pull request:
### Type of change

- Enhancement / new feature

### Description

Resolves #351

In this PR, every `FilterHandler` will detect whether the user has deferred the result (by checking if the `CompletionStage` returned by the Filter is done). If the result is deferred, that Filter will stop processing messages, instead adding them to a queue. Then when the user completes their async work, it will process that deferred result and then drain the queue, processing each of the enqueued messages.

We also worry about Client messages piling up in that queue as in the case of ProduceRequest messages with zero acks, the Client could keep sending in more data expecting no response. This could exhaust memory. This PR naively disables autoread on the Channel when we detect the user has deferred the result and re-enables it when the result is completed. In future we could implement something more sophisticated using an intermediate buffer, and only pausing when that buffer fills. For now, since this is an operation we will educate users about as being costly I think it's okay if this case is not optimised.

To avoid pushing lots of small tasks onto the netty executor (when we call channel context methods from another thread), we also make sure to switch back to the netty thread early, soon after the user completes the result. So that handling the result and then draining the enqueued work will all be done on the netty loop.

We also set up timeouts on the deferred work, hardcoded to 20 seconds the same as the extra request timeout. If the user's future isn't completed in that time then it will be completed exceptionally from the netty executor.

Note that the extra requests you can send to the broker with `contexst.sendRequest` are not subject to any queueing. We want to send them towards the broker and process the result without this queuing logic interfering.

So a detailed example of a message flow might look like:

1. Client sends request A, then request B
2. Kroxylicious Filter receives request A and goes async, sending an HTTP request with an async client
3. Kroxylicious pauses the inbound channel's autoread to stop reading messages from the client and schedules timeouts on the netty executor
4. Kroxylicious Filter receives request B, because request A was deferred it enqueues request B
5. The HTTP request completes and eventually it completes the `CompletionStage<RequestFilterResult>` forwarding an augmented request A towards the broker. The inbound channel's autoread is enabled so it can read client messages again.
6. As soon as forwarding request A is finished it continues to drain the queue, and processes request B
7. Request B is forwarded towards the broker
8. Response A from the broker is received by Kroxylicious and forwarded toward the client
9. Response B from the broker is received by Kroxylicious and forwarded toward the client
10. Client gets response A, then response B :tada: 

### Additional Context

This is continuing from where #494 left off. It made the Filter API asynchronous so that the interfaces are in place, but the delaying logic isn't build yet.

We want to support users tactically doing asynchronous work in their Filters. This should be kept to a minimum but we do want the possibility to delay forwarding while some work is being done in another thread. For example:

```
class PrefixingFilter implements RequestFilter {
    @Override
    public CompletionStage<RequestFilterResult> onRequest(ApiKeys apiKey,
                                                          RequestHeaderData header,
                                                          ApiMessage body,
                                                          KrpcFilterContext filterContext) {
       var someFuture = someExternalActionThatReturnsFuture();
       return someFuture.thenApply(r -> {
           return filterContext.forwardRequest(header, body);
       });
    }
}
```
This must be done with care because if we continue to forward messages through the pipeline after we have handed over the work to another thread, then we may respond to the Kafka Client with out-of-order messages. We want to stop reading and processing messages from the client while we are in this deferred state, and resume once they complete the stage.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Review performance test results. Ensure that any degradations to performance numbers are understood and justified.
- [ ] Make sure all Sonar-Lint warnings are addressed or are justifiably ignored.
- [ ] Update documentation
- [ ] Reference relevant issue(s) and close them after merging
- [ ] For user facing changes, update CHANGELOG.md (remember to include changes affecting the API of the test artefacts too).

robobario commented on kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java:
```diff
@@ -38,13 +42,17 @@ public class FilterHandler extends ChannelDuplexHandler {
     private final String sniHostname;
+    private final Channel inboundChannel;
     private final FilterInvoker invoker;
+    private CompletableFuture<Void> writeFuture = CompletableFuture.completedFuture(null);
```
How do you feel about queuing up Futures and chaining them together, as opposed to having something here like a `Deque<DelayedRequest>` and `Deque<DelayedResponse>` plus variables to track what state we are in, like `boolean requestDeferred`.

Using Future lets us combine those states.

k-wall replied:
I'm okay with that approach.

k-wall commented on kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java:
```diff
@@ -53,6 +61,19 @@ String filterDescriptor() {
+            doWrite(ctx, msg, promise);
+        }
+        else if (writeFuture.isDone()) {
+            writeFuture = doWrite(ctx, msg, promise);
```
to self - I suppose we are doing this to avoid a big long chain of completed futures.  it should be safe, but makes me worry we might lose track of something.

k-wall commented on kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java:
```diff
@@ -66,9 +87,12 @@ public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
+            var future = stage.toCompletableFuture();
+            var withTimeout = handleDeferredStage(ctx, future);
+            var onNettyThread = switchBackToNettyIfRequired(ctx, withTimeout);
+            return onNettyThread.whenComplete((requestFilterResult, t) -> {
                 // maybe better to run the whole thing on the netty thread.
```
```suggestion
            return onNettyThread.whenComplete((requestFilterResult, t) -> {
```

k-wall commented on kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java:
```diff
@@ -112,65 +135,121 @@ else if (requestFilterResult.message() != null) {
+    private <T extends FilterResult> CompletableFuture<T> handleDeferredStage(ChannelHandlerContext ctx, CompletableFuture<T> stage) {
+        if (!stage.isDone()) {
+            inboundChannel.config().setAutoRead(false);
+            ctx.executor().schedule(() -> {
```
do we need to cancel this?

k-wall replied:
See my PR

k-wall commented on kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java:
```diff
@@ -112,65 +135,121 @@ else if (requestFilterResult.message() != null) {
+        if (!stage.isDone()) {
+            inboundChannel.config().setAutoRead(false);
+            ctx.executor().schedule(() -> {
+                stage.completeExceptionally(new TimeoutException());
```
Might need some details of what filter stage actually timed out.

k-wall replied:
See my PR

k-wall commented on kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java:
```diff
@@ -112,65 +135,121 @@ else if (requestFilterResult.message() != null) {
+
+    private <T extends FilterResult> CompletableFuture<T> handleDeferredStage(ChannelHandlerContext ctx, CompletableFuture<T> stage) {
+        if (!stage.isDone()) {
+            inboundChannel.config().setAutoRead(false);
```
Not for now - we need metrics about how long we stop owing to awaiting the async work of filters.

robobario replied:
#520

k-wall commented on kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java:
```diff
@@ -112,65 +135,121 @@ else if (requestFilterResult.message() != null) {
+    }
+
+    private <T> CompletableFuture<T> switchBackToNettyIfRequired(ChannelHandlerContext ctx, CompletableFuture<T> future) {
+        CompletableFuture<T> nettyDrivenFuture = new CompletableFuture<>();
```
I'm not grasping why completing the future has to be done on the netty thread.

robobario replied:
Might drop this commit as being surplus to requirement and come back to it. 

I was speculating it could be kinder to netty to switch back once instead of making multiple calls to the netty context as we drain the queue, each call getting put into the executor queue. But that needs testing.

Maybe a better argument is it could make increase our confident that the work of the FilterHandler is executed single-threaded (though the queue should enforce that), and puts the framework in control of where kroxylicious-owned logic is executed.

SamBarker replied:
I think splitting this change out is the right call. As I'm not entirely sure `ctx.executor().execute()` is going to take us back to the originating thread or if its just going to cause more context thrashing. It probably warrants a benchmark (we can craft one)

robobario replied:
Reverted the commit, so it's removed from this PR :+1:

k-wall commented on kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java:
```diff
@@ -112,65 +135,121 @@ else if (requestFilterResult.message() != null) {
+            readFuture = doRead(ctx, msg);
+        }
+        else {
+            readFuture = readFuture.whenComplete((a, b) -> doRead(ctx, msg));
```
are we swallowing an error (case where b != null)

k-wall replied:
I've added unit tests and push fixes to my branch.

k-wall commented on integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java:
```diff
@@ -164,6 +166,29 @@ public void requestFiltersCanRespondWithoutProxying(KafkaCluster cluster, Admin
     }
 
+    @ParameterizedTest
+    @ValueSource(strings = { "RequestForwardDelaying", "ResponseForwardDelaying" })
```
I've covered that in my PR to yours

robobario commented:
@strimzi-ci run perf

SamBarker commented on integrationtests/src/test/java/io/kroxylicious/proxy/filter/RequestForwardDelayingFilter.java:
```diff
@@ -0,0 +1,32 @@
+    public CompletionStage<RequestFilterResult> onRequest(ApiKeys apiKey, RequestHeaderData header, ApiMessage body, KrpcFilterContext filterContext) {
+        CompletableFuture<RequestFilterResult> future = new CompletableFuture<>();
+        var builder = filterContext.requestFilterResultBuilder();
+        try (var executor = Executors.newScheduledThreadPool(1)) {
```
Hmmm maybe this is the case for per filter instance life cycles. Then again why would each filter instance need an isolated executor? Or at least one that wasn't supplied by a contributor.

robobario replied:
it just wants a way to delay, so I imagine running that on the netty executor would be fine. Making that available via the Contributor/Constructor would be great.

k-wall replied:
In this test-use case, I'm happy to see the filter manage its own thread pool.     The wider point is valid: filters ought to have a way to share a thread pool.  Presumable, the share state feature we are imagining would permit this.

robobario replied:
True, for this test we do want to be certain it's working off the netty thread.

SamBarker replied:
Yeah, I was thinking of the wider point rather than this specific instance (which is fine).

I am expecting the shared state or the injection context API's to offer a solution something similar to the one @robobario outlined but I'm still not quite sure which or where.

SamBarker commented on kroxylicious-test-tools/src/test/java/io/kroxylicious/test/tester/KroxyliciousTestersTest.java:
```diff
@@ -216,9 +216,15 @@ private static Map<String, Object> randomGroupIdAndEarliestReset() {
+            assertEquals(1, records.count());
+        }
+        finally {
+            consumer.close();
```
A bit of a surprising side effect of calling `assertOneRecordConsumedFrom`. 

Also isn't it considered bad form to close a resource someone else allocated?

robobario replied:
Reworked it a little to create-close the consumer with a try-with-resources, the test was also @Disabled with a comment about it not working from Keith, was this resolved with the changes around flushing @k-wall? The test passes now at least on my machine.

k-wall replied:
@robobario "was this resolved with the changes around flushing" yes.  I'd forgotten - thank you.
@SamBarker assertOneRecordConsumedFrom - "close a resource someone else allocated" <- dumb change on my behalf. @robobario's refactor looks sensible!

SamBarker commented on kroxylicious/src/test/java/io/kroxylicious/proxy/internal/FilterHandlerTest.java:
```diff
@@ -52,6 +63,150 @@ public void testForwardRequest() {
+        assertThat(channel.isOpen()).isFalse();
+        var propagated = channel.readInbound();
+        assertThat(propagated).isNull();
+        assertThat(seen).hasSize(1);
+        assertThat(seen).containsExactly(frame1.body());
```
One can chain these 
```suggestion
        assertThat(seen).hasSize(1).containsExactly(frame1.body());
```

Though when using `containsExactly` does adding a `size == 1` assertion add anything?

SamBarker commented on kroxylicious/src/test/java/io/kroxylicious/proxy/internal/FilterHandlerTest.java:
```diff
@@ -52,6 +63,150 @@ public void testForwardRequest() {
+    }
+
+    static Stream<Arguments> responseFilterClosesChannel() {
+        return Stream.of(
```
Should we be covering `ShortCircuit` and `Drop` here as well?

robobario replied:
drop doesn't combine with close (I assume because drop+close has the same effect as close).

but yes I think we could have a test for a short-circuit response + close (though it's a request test, so the other `Stream<Arguments>` method)

SamBarker replied:
Both good points...

k-wall replied:
I added a unit test to my PR `#requestShortCircuit`

SamBarker commented on kroxylicious/src/test/java/io/kroxylicious/proxy/internal/FilterHandlerTest.java:
```diff
@@ -52,6 +63,150 @@ public void testForwardRequest() {
+        ApiVersionsRequestFilter filter = (apiVersion, header, request, context) -> filterFuture;
+        buildChannel(filter, 50L);
+        writeRequest(new ApiVersionsRequestData());
+        Thread.sleep(60L);
```
Can we wait for something to happen instead of just sleeping?

robobario replied:
Yes, it looks like embedded channel has `freezeTime` and `advanceTime` type methods on it.

SamBarker commented on kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java:
```diff
@@ -112,65 +143,133 @@ else if (requestFilterResult.message() != null) {
+                doRead(ctx, msg);
+            }
+        }
+        else if (readFuture.isDone()) {
```
Should we care about successful vs unsuccessful completion here?

k-wall replied:
At the moment, I don't think we need to care. By my reading of the FilterHander, those future will only ever be completed successfully, even in the cases where the connection is closed,  there is a filter ends exceptionally or a filter unexpected returns null.

Maybe something in motivate us to get more sophisticated later, but I think what we have is already a valuable step forward.

SamBarker reviewed (commented):
I think this makes sense. I'm minded to approve but I want to take another pass at `doRead` once the switch to the netty executor is factored out.

robobario commented:
Thanks @SamBarker will look more into your comments tomorrow, @k-wall if you have any time feel free to make changes.

robobario commented on api/kroxylicious-api/kroxylicious-api.iml:
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
I don't know, maybe it happened when I switched branches to v0.2.0 to do topic-encryption demos. Also, I have copious changes I'm avoiding committing after installing the sonarlint plugin. Need to revisit which IDEA files need checking in.

SamBarker replied:
Its not just happening for you. Its happened to me locally and on another PR I've seen.

k-wall commented on integrationtests/src/test/java/io/kroxylicious/proxy/KrpcFilterIT.java:
```diff
@@ -164,6 +168,66 @@ public void requestFiltersCanRespondWithoutProxying(KafkaCluster cluster, Admin
+            // The LIST_TRANSACTIONS RPC is the trigger to the filter to make the async request. The filter
+            // won't send the LIST_TRANSACTIONS response to us until the async request completes. We
+            // don't actually care about the content of the response, only that it returns,
+            admin.listTransactions().all().get(10, TimeUnit.SECONDS);
```
The idea of the test was to make sure that, if the async work done by the forward is itself work done against the cluster down the same connection, that requests/responses still flowed correctly.   That's the key difference between this test and `supportsRequestResponseForwardAsynchronicity`.

The important part of the test is that fact that list transaction response is received.   The defect was shown by the fact that it wasn't (the message was stuck in a Netty buffer)

The publishing of the message by the async task and the consuming of it by the test was a way to verify that the async work had actually happened.  

> I also wonder about polling the topic before list transactions returns? Something along the lines of

The objective of the test wasn't to show that Kroxylicious doesn't somehow, return a response prematurely (which would be a coding error in the filter).

I'll have a think a making the test more simple. I really just need a way to mark the response, so the filter can demonstrate conclusively that it did await the async work. I'll take a look at mock server.

k-wall replied:
I've refactored the test on this PR https://github.com/robobario/kproxy/pull/7 made against this one.

k-wall reviewed (approved):
I think this work is  in good shape.
We need to refactor FilterHandler/DefaultFilterContext to have better division of responsibilities, but maybe that's a job for the next PR.   I think what we have here a great step forward.

robobario commented:
Planning to merge with some [outstanding comments](https://github.com/robobario/kproxy/pull/7#discussion_r1286444674) about further simplifications to the IT. We can revisit them later.

robobario commented:
@k-wall I found that `SampleFilterIntegrationTest` was failing intermittently on my machine. 

A minimal non-deterministic reproducer was:
```
@Test
    public void testIt(KafkaCluster cluster) {
        try (var tester = kroxyliciousTester(proxy(cluster)); Admin admin = tester.admin(); var producer = tester.producer()) {
            admin.createTopics(List.of(new NewTopic("topic", 1, (short) 1))).all().get(10, TimeUnit.SECONDS);
        }
        catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
```
It requires the unused producer to trigger the failure, something to do with there being multiple metadata responses in flight, or multiple event loops with channels assigned.

In the end I found that the `RequestResponseMarkingFilter` wasn't exercising the incomplete future path because the `close()` of the executor was blocking until it had finished doing it's work. So the Future was complete by the time it got back to `FilterHandler`. Once I changed this to not block by calling `shutdown()` instead of close `supportsForwardDeferredByAsynchronousAction` failed on the response path. I fixed it by adding in a flush to the inbound channel after a deferred read.

I think we might be missing a test for an asynchronous short-circuit response. Currently I've made it flush both inbound and outbound channels after a deferred write in case they short-circuit responded.

SamBarker commented on integrationtests/src/test/java/io/kroxylicious/proxy/filter/RequestResponseMarkingFilter.java:
```diff
@@ -58,11 +59,14 @@ public CompletionStage<ApiMessage> apply(KrpcFilterContext context, ApiMessage b
                 CompletableFuture<ApiMessage> result = new CompletableFuture<>();
-                try (var executor = Executors.newScheduledThreadPool(1)) {
-                    var delay = (long) (Math.random() * 200);
+                ScheduledExecutorService executor = Executors.newScheduledThreadPool(1);
```
Does the change from try-with-resources to try finally actually make a difference?

robobario replied:
yes, in finally we are calling `shutdown`:

```
Initiates an orderly shutdown in which previously submitted tasks are executed, but no new tasks will be accepted. Invocation has no additional effect if already shut down.
This method does not wait for previously submitted tasks to complete execution. Use awaitTermination to do that.
```

where close does:

```
     * Initiates an orderly shutdown in which previously submitted tasks are
     * executed, but no new tasks will be accepted. This method waits until all
     * tasks have completed execution and the executor has terminated.
```

k-wall replied:
Good catch.

SamBarker commented on kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java:
```diff
@@ -95,44 +97,11 @@ private CompletableFuture<Void> doWrite(ChannelHandlerContext ctx, Object msg, C
-            }).toCompletableFuture().thenApply(filterResult -> null);
-
+            boolean defer = !future.isDone();
+            var maybeDeferred = defer ? handleDeferredStage(ctx, future, decodedFrame) : future;
```
I'm not 100% sure we can hoist the `isDone` check out of `handleDeferredStage` as the future could complete between this check and `handleDeferredStage` executing. 

I guess its just an optimisation (we avoid creating the timeout future) so it is probably safe.

k-wall replied:
I believe the intent is to optimise a fast path for filters that return completed futures.  I think the code is meeting that aim.

SamBarker commented on kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java:
```diff
@@ -146,28 +115,59 @@ else if (requestFilterResult.message() != null) {
         }
     }
 
+    private CompletableFuture<RequestFilterResult> executeWrite(ChannelHandlerContext ctx, DecodedRequestFrame<?> decodedFrame,
+                                                                DefaultFilterContext filterContext,
+                                                                CompletableFuture<RequestFilterResult> withTimeout) {
+        return withTimeout.whenComplete((requestFilterResult, t) -> {
```
nit:
```suggestion
    private CompletableFuture<RequestFilterResult> executeWrite(ChannelHandlerContext ctx, DecodedRequestFrame<?> decodedFrame,
                                                                DefaultFilterContext filterContext,
                                                                CompletableFuture<RequestFilterResult> pendingStage) {
        return pendingStage.whenComplete((requestFilterResult, t) -> {
```

k-wall replied:
+1 to a rename.   From the point of view of this method, it doesn't know if the stage is deferred or not, so I think just plain `stage` would be sufficient.

k-wall replied:
I included it in https://github.com/robobario/kproxy/pull/8

SamBarker commented on kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java:
```diff
@@ -146,28 +115,59 @@ else if (requestFilterResult.message() != null) {
+                    var header = requestFilterResult.header() == null ? decodedFrame.header() : requestFilterResult.header();
+                    filterContext.forwardRequestInternal((RequestHeaderData) header, requestFilterResult.message());
+                }
+            }
```
Appreciate this is probably just a code shuffle, but are we missing an `else` clause here?

k-wall replied:
It is a consequence of the valid combinations of drop/close/forward.  I think the code could be expressed slightly better. Let me put up a PR.

k-wall replied:
https://github.com/robobario/kproxy/pull/8

k-wall commented on kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java:
```diff
@@ -112,66 +111,163 @@ else if (requestFilterResult.message() != null) {
         }
     }
 
+    private CompletableFuture<?> handleUndeferRead(CompletableFuture<?> execute) {
```
I'm not sure about the naming. The method `handleUndefer` had me scratching my head. How about `handleDeferredStageCompletion`

k-wall commented on kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java:
```diff
@@ -112,65 +142,108 @@ else if (requestFilterResult.message() != null) {
+                LOGGER.error("{}: Filter{} for {} response unexpectedly returned null. This is a coding error in the filter. Closing connection.",
+                        ctx.channel(), filterDescriptor(), decodedFrame.apiKey());
+                filterContext.closeConnection();
+                return CompletableFuture.completedFuture(null);
```
I don't think so.

IMO this will be a filter author programming issue (at least, it will be once we make the javadoc clear that the the filter MUST return a CompletionStage / @NoneNull applied to the methods).   I'm actually questioning whether the error should be actually a warn.

I think we will probably want metrics to count cases where filters:

* return null 
* complete exceptionally
* filter timeouts

Obviously all work for another PR.

k-wall commented:
> I think we might be missing a test for an asynchronous short-circuit response. Currently I've made it flush both inbound and outbound channels after a deferred write in case they short-circuit responded.

Agreed.  I also noticed that we lacked integrations tests that validate the a error response is received in the case where the connection is closed.  I'd actually started work on this yesterday, but realise that the mock will need an enhancement.

I think we should uncouple the need for this integration test from this PR.  I'd really like to see this PR land!

robobario commented on integrationtests/src/test/java/io/kroxylicious/proxy/filter/RequestForwardDelayingFilter.java:
```diff
@@ -0,0 +1,32 @@
+        CompletableFuture<RequestFilterResult> future = new CompletableFuture<>();
+        var builder = filterContext.requestFilterResultBuilder();
+        try (var executor = Executors.newScheduledThreadPool(1)) {
+            var delay = (long) (Math.random() * 200);
```
I made it a fixed 200 while changing how we shut it down.
